frontend/app/components/wp-panels/watchers-panel/watchers-panel.controller.ts

```diff
@@ -28,120 +28,107 @@
+  public autocompleteInput = '';
   public error = false;
   public allowedToAdd = false;
   public allowedToRemove = false;
 
-  public available:any[] = [];
-  public watching:any[] = [];
-  public text:any;
+  public watching: any[] = [];
+  public text: any;
 
   constructor(public $scope,
+              public $q,
               public I18n,
-              public wpCacheService:WorkPackageCacheService,
-              public wpWatchers) {
+              public wpNotificationsService: WorkPackageNotificationService,
+              public wpCacheService: WorkPackageCacheService) {
 
     this.text = {
       loading: I18n.t('js.watchers.label_loading'),
-      loadingError: I18n.t('js.watchers.label_error_loading')
+      loadingError: I18n.t('js.watchers.label_error_loading'),
+      autocomplete: {
+        placeholder: I18n.t('js.watchers.label_click_to_add')
+      }
     };
 
-
-    $scope.$on('watchers.add', (evt, watcher) => {
-      this.addWatcher(evt, watcher);
-    });
-    $scope.$on('watchers.remove', (evt, watcher) => {
-      this.removeWatcher(evt, watcher);
-    });
-
     if (!this.workPackage) {
       return;
     }
 
     wpCacheService.loadWorkPackage(<number> this.workPackage.id).observe($scope)
-      .subscribe((wp:WorkPackageResourceInterface) => {
+      .subscribe((wp: WorkPackageResourceInterface) => {
         this.workPackage = wp;
-        this.fetchWatchers();
+        this.loadCurrentWatchers();
       });
   }
 
-  public fetchWatchers() {
+  public loadCurrentWatchers() {
     this.error = false;
     this.allowedToAdd = !!this.workPackage.addWatcher;
     this.allowedToRemove = !!this.workPackage.removeWatcher;
 
-    this.wpWatchers.forWorkPackage(this.workPackage)
-      .then(users => {
-        this.watching = users.watching;
-        this.available = users.available;
+    this.workPackage.watchers.$load()
+      .then((collection:CollectionResource) => {
+        this.watching = collection.elements;
       })
-      .catch(() => {
-        this.watching = [];
-        this.available = [];
-        this.error = true;
+      .catch((error) => {
+        this.wpNotificationsService.showError(error, this.workPackage);
       });
   };
 
-  public addWatcher(event, watcher) {
-    event.stopPropagation();
-
-    watcher.loading = true;
-    this.add(watcher, this.watching);
-    this.remove(watcher, this.available);
+  public autocompleteWatchers(query) {
+    if (!query) {
+      return [];
+    }
 
-    this.loadingPromise = this.wpWatchers.addForWorkPackage(this.workPackage, watcher)
-      .then(watcher => {
-        this.$scope.$broadcast('watchers.add.finished', watcher);
+    const deferred = this.$q.defer();
+    this.autocompleteLoadingPromise = deferred.promise;
+
+    this.workPackage.availableWatchers.$link.$fetch(
+      {
+        filters: JSON.stringify([{
+          name: {
+            operator: '~',
+            values: query,
+          }
+        }]),
```

Comment:
I think sorting by name would be nice as well:

![image](https://cloud.githubusercontent.com/assets/617519/20309032/b466bd8e-ab46-11e6-96a9-b2ff77253034.png)